api/replica.go

```diff
@@ -14,20 +19,66 @@ type Replica struct {
 	handler   http.Handler
 	tlsConfig *tls.Config
 	primary   string
+	addr      string // the address of this replica
+	// mu is the mutex used for locking the primary.
+	mu sync.RWMutex
+	// pWait is used to block execution of ServeHTTP until the primary is ready
+	// primaryWait is only used by readers, and is populated by the output of
+	// mu.RLocker, which means the lock associated with it is actually a read
+	// lock.
+	primaryWait *sync.Cond
+
+	// capture the hijack function for easier unit testing
+	hijack func(*tls.Config, string, http.ResponseWriter, *http.Request) error
 }
 
 // NewReplica creates a new API replica.
-func NewReplica(handler http.Handler, tlsConfig *tls.Config) *Replica {
-	return &Replica{
+func NewReplica(handler http.Handler, tlsConfig *tls.Config, addr string) *Replica {
+	r := &Replica{
 		handler:   handler,
 		tlsConfig: tlsConfig,
+		addr:      addr,
+		hijack:    hijack,
 	}
+	r.primaryWait = sync.NewCond(r.mu.RLocker())
+	return r
 }
 
 // SetPrimary sets the address of the primary Swarm manager
 func (p *Replica) SetPrimary(primary string) {
 	// FIXME: We have to kill current connections before doing this.
+	// this lock on the mutex is a write lock so that we can write the primary
+	// out safely
+	p.mu.Lock()
+	defer p.mu.Unlock()
 	p.primary = primary
+	if p.primary != "" {
+		// broadcast to wake everyone waiting on a nonempty primary
+		p.primaryWait.Broadcast()
+	}
+}
+
+// getPrimary returns the primary if it exists or an error if the context
```

Comment:
nit: getPrimary returns the primary if it exists or block until the context is done.